### DISCUSSION TITLE
Using upytorch as a library for newest esp32-s3(Flash:16MB-RAM:8MB)

### DISCUSSION THREAD

**Discussion by smartengm** [micropython/micropython#14306] [ESP32]
Dear experts,

I am working with micropython(Version:1.19) and ESP32-S3(MON16R8) and I want to use a light version of Pytorch for this microcontroller. After searching on Internet, I found this URL:  

https://github.com/ljk53/upytorch

Here are my questions:

1-Can we use upytorch as a library for micropython(1.19) in Thonny editor? How?...Is upytorch  compatible to the new ESP32-S3 series?
(Thonny "manage packages" can not find upytorch as a standard library.)

2-It seems that we have to create a new firmware of micropython for ESP32 to work with upytorch. Am I right? How can we create this firmware by windows 10 and for the new  ESP32-S3 ?Is upytorch compatible to micropython(version:1.19) or the latest version of  micropython?

3- How close is upytorch to the real Pytorch package in python? Is upytorch capable enough to implement light deep learning models on this microcontroller?  

Your consideration is highly appreciated.
Regards:
smartengm.

**Comment by jonnor** [comment]
These questions are probably best asked on that repository. People that just know MicroPython generally are unlikely to know these detailed kind of answers about a third party library.

There are instructions in that repository for how to build a custom firmware for ESP32. This might or might not work on ESP-S3 - best bet is just to try to get it to work. And yes, I believe a custom build is the way to set it up. It does not seem possible to install it in any other way.

**Comment by jonnor** [comment]
An alternative ML library is https://github.com/emlearn/emlearn-micropython/ (disclosure: I am the maintainer). It is designed to be easy to install (just copy a file). Support for Convolutional Neural Networks is coming, as well as simple MLPs.

**Reply by smartengm** [reply]
Hi @jonnor,

Thank you for your reply. I think I know about *emlearn*  ,but not completely .
Is it possible to use your MLPs to implement reinforcement learning on microcontrollers? Specially for the training phase?

Thanks again.

**Reply by jonnor** [reply]
Hi @smartengm - no emlearn focuses on inference, so there is no training code implemented. On-device training for microcontrollers is generally even rarer than on-device inference. I would not expect to find a finished framework - be prepared to write a lot from scratch. But one finds some experiments lying around that might be reasonable starting points. Example, https://github.com/wozeparrot/tinygrad-on-esp32
